app/src/main/java/com/rajatgoyal/puzzle15/util/AchievementHandler.java

```diff
@@ -0,0 +1,116 @@
+package com.rajatgoyal.puzzle15.util;
+
+import android.app.Activity;
+import android.widget.Toast;
+
+import com.google.android.gms.games.AchievementsClient;
+import com.rajatgoyal.puzzle15.R;
+
+import timber.log.Timber;
+
+/**
+ * Created by rajatgoyal715 on 20/2/19.
+ */
+public class AchievementHandler {
+
+    private AchievementsClient client;
+
+    private static int[] PLAYED_IDS = {
+            R.string.achievement_newbie,
+    };
+    private static int[] COMPLETED_IDS = {
+            R.string.achievement_novice__1_game,
+            R.string.achievement_beginner__5_games,
+            R.string.achievement_skilled__10_games,
+            R.string.achievement_proficient__15_games,
+            R.string.achievement_experienced__20_games,
+            R.string.achievement_advanced__30_games,
+            R.string.achievement_expert__50_games,
+            R.string.achievement_master__80_games,
+            R.string.achievement_legend__100_games
+    };
+
+    private static int[] TIMES = {600, 300, 120, 90, 60};
+    private static int[] TIME_IDS = {
+            R.string.achievement_snail__10_min,
+            R.string.achievement_tortoise__5_min,
+            R.string.achievement_rabbit__2_min,
+            R.string.achievement_deer__1_min_30_secs,
+            R.string.achievement_cheetah__1_min,
+    };
+
+
+    private static int[] MOVES = {300, 250, 200, 150, 100};
+    private static int[] MOVES_IDS = {
+            R.string.achievement_wood__300_moves,
+            R.string.achievement_copper__250_moves,
+            R.string.achievement_silver__200_moves,
+            R.string.achievement_gold__150_moves,
+            R.string.achievement_diamond__100_moves,
+    };
+
+    public AchievementHandler(AchievementsClient achievementsClient) {
+        this.client = achievementsClient;
+    }
+
+    private void unlockOnUIThread(final Activity activity, final int achievementId) {
+        if (this.client == null) {
+            Toast.makeText(activity, "Achievement Client is null.", Toast.LENGTH_SHORT).show();
+            Timber.d("Please login first");
+            return;
+        }
+        final AchievementsClient achievementsClient = this.client;
+        activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                achievementsClient.unlock(activity.getString(achievementId));
+            }
+        });
+    }
+
+    private void incrementAchievement(final Activity activity, final int achievementId) {
```

Comment:
Why would you pass whole activity to these method, just pass the context, this is a very bad practice

Reply:
I had to access runOnUiThread and getString methods using Activity instance. I see that we can typecast the context to Activity, but I don't recall why I didn't do that earlier. I'll try using context now.

Reply:
try that, even if that doesn't work, avoid sending activity to this method, instead use interfaces, this could potentially cause memory leak